### DISCUSSION TITLE
Real translations

### DISCUSSION THREAD

**Discussion by trash-anger** [forwardemail/forwardemail.net#205] [Ideas]
Hey,
I just discovered your service and I have to say that I'm amazed by it! Thank you very much!!!! :)
I may have a recommandation, the auto translate is absolutely useless as the translations are completely fuck-up (for french at least)...
You know where to find me if you need good french translations! And once again, thank you!!!

**Comment by titanism** [comment]
@trash-anger hey there 👋  our service uses <https://github.com/ladjs/mandarin> under the hood (which we also made) &ndash; it's not perfect as it doesn't get the full context of phrases, but it breaks them apart according to HTML tags, paragraphs, etc.

The locales are in the `locales` folder at <https://github.com/forwardemail/forwardemail.net/tree/master/locales>, for example French is at https://github.com/forwardemail/forwardemail.net/blob/master/locales/fr.json

If you want to contribute, you could click the pencil icon and edit it, or fork locally and submit a pull request.  Happy to payout a bounty over PayPal or GitHub sponsors depending on the extent of effort!

**Reply by trash-anger** [reply]
I'll send a PR soon.

**Comment by shaunwarman** [comment]
@trash-anger I'm also curious your thoughts on the translation differences using something like gpt-4.

| english | french (google) | french (gpt-4) |
|--------|-----------------|----------------|
| Domain name was invalid (must be an IP or FQDN). | Le nom de domaine n'était pas valide (doit être une adresse IP ou un nom de domaine complet). | Le nom de domaine n'était pas valide (doit être une IP ou un FQDN). |
| Domain name was invalid (must be a FQDN). | Le nom de domaine n'était pas valide (doit être un nom de domaine complet). | Le nom de domaine n'était pas valide (doit être un FQDN). |
| You have successfully accepted the invite to this domain and you can now create your first alias. | Vous avez accepté avec succès l'invitation à ce domaine et vous pouvez maintenant créer votre premier alias. | Vous avez accepté avec succès l'invitation à ce domaine et vous pouvez maintenant créer votre premier alias. |
| An error occurred while emailing the invite link.  Please manually copy the the invite link and share it with this person if needed. | Une erreur s'est produite lors de l'envoi par e-mail du lien d'invitation. Veuillez copier manuellement le lien d'invitation et le partager avec cette personne si nécessaire. | Une erreur s'est produite lors de l'envoi du lien d'invitation par courriel. Veuillez copier manuellement le lien d'invitation et le partager avec cette personne si nécessaire. |
| Group was invalid, must be either admin or user. | Le groupe n'était pas valide, doit être administrateur ou utilisateur. | Le groupe était invalide, il doit être soit admin soit utilisateur. |

**Reply by trash-anger** [reply]
There's no big difference here... I'm able to understand both. I think there's some issues when there's no context around what has to be translated. ([cf this answer](https://github.com/orgs/forwardemail/discussions/205#discussioncomment-7824411))

But the google results are slightly better.

**Comment by titanism** [comment]
@trash-anger Can you please share at least one example of where the translations are off?  What are they now and what should the desired translation actually be

**Reply by titanism** [reply]
We are exploring using some other options (e.g. GPT-4) so we have contextual translations with raw markdown conversion instead of tokenization translation approach.  We were curious if you had an example you could share from our FAQ or elsewhere on our website with what the translation is currently and what it should be (e.g. why is it off?).

**Reply by trash-anger** [reply]
I guess there's sometime not enough context for the translation to be good.
Here's some examples:
[Les soutiens](https://github.com/forwardemail/forwardemail.net/blob/f25db63b790cedf138be8d3b757b39c424c95a32/locales/fr.json#L5352) should be "Supporte" ("Les soutiens" means "supporters").
[Des guides](https://github.com/forwardemail/forwardemail.net/blob/f25db63b790cedf138be8d3b757b39c424c95a32/locales/fr.json#L124) should be "Guides" ("Des guides" means "Some guides" which is really weird for a header link)
[Cloudflare](https://github.com/forwardemail/forwardemail.net/blob/f25db63b790cedf138be8d3b757b39c424c95a32/locales/fr.json#L901C14-L901C14) should be Cloudflare ("Nuageux" is very funny but we call CloudFlare cloudflare, even in french)

If you still want to keep you gpt-4 translation pipeline, we could duplicate the local and override the gpt-4 generated by one manually translated... what do you think of that approach?

**Reply by titanism** [reply]
Hi there - we do support manual translation overrides right now, which means we can accept PR's at anytime for any locale.  The way it works is by checking if `"foo": "foo"`, if the key matches the value (and it is not English) then it is assumed to be not translated, so then our automated translation with `mandarin` will take place.

**Reply by titanism** [reply]
@trash-anger we've fixed `Les soutiens`, `Des guides`, and `Cloudflare` and will be pushing and deploying shortly.  See changes to `locales/fr.json` in this repo soon.
